### PR TITLE
SGA-6994 Adding satori attributes as a resource to our TerraformProvider.

### PR DESCRIPTION
**Pull request by ron-ts**

### CODE REVIEW

File: GNUmakefile [hunk 1]
```diff
@@ -5,6 +5,7 @@ VERSION=1.0.0
 BINARY=terraform-provider-${NAME}
 OS_ARCH=darwin_amd64
 # Switch between the two OS architecture according to own computer's architecture when working locally (if needed)
+# arm64 is the OS_ARCH for m1 or m2 apple processor
```
Comment (otlg) [inline GNUmakefile:8]: Put this, let's resolve it dynamically 
`
UNAME_P := $(shell uname -p)
ifeq ($(filter %86,$(UNAME_P)),)
	OS_ARCH=darwin_amd64
endif
ifneq ($(filter arm%,$(UNAME_P)),)
	OS_ARCH=darwin_arm64
endif
$(info Current architecture (OS_ARCH) is $(OS_ARCH))
`

Comment (otlg) [reply GNUmakefile:8]: UNAME_P := $(shell uname -p)
ifeq ($(filter %86,$(UNAME_P)),)
	OS_ARCH=darwin_amd64
endif
ifneq ($(filter arm%,$(UNAME_P)),)
	OS_ARCH=darwin_arm64
endif
$(info Current architecture (OS_ARCH) is $(OS_ARCH))

File: satori/resources/resource_user_custom_attributes_common.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
+package resources
+
+import (
+	"encoding/json"
+	"fmt"
```
Comment (otlg) [inline satori/resources/resource_user_custom_attributes_common.go:5]: Please check your editor indentation, it's too big, it should be 2 spaces...

File: docs/resources/user_attributes.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
+---
+# generated by https://github.com/hashicorp/terraform-plugin-docs
+page_title: "satori_user_attributes Resource - terraform-provider-satori"
+subcategory: ""
+description: |-
+  User Satori Attributes.
+---
+
+# satori_user_attributes (Resource)
```
Comment (otlg) [inline docs/resources/user_attributes.md:9]: What do you think if we will call the resource satori_user? 
In future we can add more staff there....

File: satori/resources/utils.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package resources
+
+import (
+	"encoding/json"
+	"os"
+)
+
+func normalizeDataJSON(val interface{}) string {
+	dataMap := map[string]interface{}{}
+	fileContent, readFileErr := os.ReadFile("/Users/rontzabary/workspace/local-provider-scripts/test.json")
```
Comment (otlg) [inline satori/resources/utils.go:10]: `/Users/rontzabary/workspace/local-provider-scripts/test.json` ?

File: satori/resources/utils.go [hunk 5]
```diff
@@ -0,0 +1,21 @@
+package resources
+
+import (
+	"encoding/json"
+	"os"
+)
+
+func normalizeDataJSON(val interface{}) string {
```
Comment (otlg) [inline satori/resources/utils.go:8]: What does this function do?

File: satori/api/users.go [hunk 6]
```diff
@@ -1,16 +1,75 @@
...
 
 type User struct {
-	Id    string `json:"id"`
-	Email string `json:"email"`
+	Id         string            `json:"id"`
+	Email      string            `json:"email"`
+	Attributes map[string]string `json:"customAttributes,omitempty"`
+}
+
+type UserWithSatoriAttributes struct {
```
Comment (otlg) [inline satori/api/users.go:15]: ???

File: docs/resources/user_attributes.md (hunk 3 above)
Comment (otlg) [reply docs/resources/user_attributes.md:9]: satori_user_setting

File: satori/resources/resource_user_custom_attributes.go [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+
+	c := m.(*api.Client)
+
+	result, err := createUserAttributes(d, c)
+
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	fmt.Println("In the CREATE function with response result --->", result)
```
Comment (loren-tsahalon) [inline satori/resources/resource_user_custom_attributes.go:39]: Remove this log

File: satori/resources/resource_user_custom_attributes.go [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+	if d == nil {
+		return nil, nil
+	}
+
+	attrRawJson := d.Get(Attributes).(string)
+	attrDto.UserId = d.Get(UserId).(string)
+
+	fileContent, readFileErr := os.ReadFile(attrRawJson)
+
+	if readFileErr == nil {
```
Comment (loren-tsahalon) [inline satori/resources/resource_user_custom_attributes.go:72]: What happens if error?

File: satori/resources/resource_user_custom_attributes.go [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+
+func resourceUserCustomAttributesRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	var diags diag.Diagnostics
+	c := m.(*api.Client)
+
+	res, err := getUserAttributes(c, d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	fmt.Println(res)
```
Comment (loren-tsahalon) [inline satori/resources/resource_user_custom_attributes.go:98]: remove log

File: satori/resources/resource_user_custom_attributes.go [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+
+	attr := api.UserWithCustomAttributes{}
+
+	attr.UserId = d.Get(UserId).(string)
+	attr.CustomAttributes = mergedAttributes
+
+	return &attr, err
+}
+
+// This function merges the attributes from the
```
Comment (loren-tsahalon) [inline satori/resources/resource_user_custom_attributes.go:145]: Complete the description sentence please

File: satori/resources/resource_user_custom_attributes.go [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+func resourceUserCustomAttributesUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	c := m.(*api.Client)
+
+	result, err := updateSatoriAttributes(d, c)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	fmt.Println(result)
```
Comment (loren-tsahalon) [inline satori/resources/resource_user_custom_attributes.go:190]: remove log

File: satori/resources/resource_user_custom_attributes.go [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+	var diags diag.Diagnostics
+
+	c := m.(*api.Client)
+
+	result, err := deleteUserSatoriAttributes(d, c)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	fmt.Println(result)
```
Comment (loren-tsahalon) [inline satori/resources/resource_user_custom_attributes.go:215]: remove log

File: satori/resources/resource_user_custom_attributes.go [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+}
+
+func deleteUserSatoriAttributes(d *schema.ResourceData, c *api.Client) (*api.UserWithCustomAttributes, error) {
+	input := api.UserWithCustomAttributes{}
+	input.UserId = d.Get(UserId).(string)
+
+	result, err := c.DeleteUserCustomAttributes(&input)
+
+	// Setting terraform-id from the empty string ---> triggers delete
+	d.SetId("")
```
Comment (loren-tsahalon) [inline satori/resources/resource_user_custom_attributes.go:227]: Why do you need this? didn't understand

File: satori/resources/resource_user_custom_attributes_common.go [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+			ValidateDiagFunc: func(i interface{}, k cty.Path) diag.Diagnostics {
+				var unparsedJsonString map[string]interface{}
+
+				v, ok := i.(string)
+
+				if !ok {
+					return diag.Errorf("Expected type %s to be string", k)
+				}
+
+				fileContent, readFileErr := os.ReadFile(v)
```
[the remark this answers is not in the document]
Comment (ron-ts) [reply satori/resources/resource_user_custom_attributes_common.go:39]: readFileErr == nil and it keeps working with the JSON raw data that the user entered

File: satori/resources/resource_user_custom_attributes_common.go [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+
+				if readFileErr == nil {
```
[the remark this answers is not in the document]
Comment (ron-ts) [reply satori/resources/resource_user_custom_attributes_common.go:41]: If there is an error then the resource infer that it is not a file and it uses the entered value as a json raw string

File: satori/resources/resource_user_custom_attributes_common.go [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+					v = string(fileContent)
+				}
+
+				// Validate that the user enters a map format and not just a json node e.g. string/number/boolean...
+				err := json.Unmarshal([]byte(v), &unparsedJsonString)
+
+				if err != nil {
+					return diag.Errorf("Failed to parse rawJSON string at path %s, with err %s", k, err)
+				}
+				// TODO: change the validation method to the appropriate 1
```
[the remark this answers is not in the document]
Comment (ron-ts) [reply satori/resources/resource_user_custom_attributes_common.go:51]: Nope, removing

File: satori/resources/resource_user_custom_attributes.go (hunk 13 above)
Comment (ron-ts) [reply satori/resources/resource_user_custom_attributes.go:227]: d.SetId("") triggers the terraform's object deletion from the terraform.state

File: go.mod [hunk 17]
```diff
@@ -3,6 +3,7 @@ module github.com/satoricyber/terraform-provider-satori
 go 1.16
 
 require (
+	github.com/hashicorp/go-cty v1.4.1-0.20200414143053-d3edf31b6320 // indirect
```
[the remark this answers is not in the document]
Comment (ron-ts) [reply go.mod:6]: For paths inside an object.
In the ValidationDiagFunction.

Screen Shot 2023-05-07 at 19 12 11


For more informative errors

[the remark this answers is not in the document]
Comment (ron-ts) [reply go.mod:6]: Note the signature and the ```diag.Errorf(..., k, err)``` formatting

File: docs/resources/user_attributes.md [hunk 18]
```diff
@@ -0,0 +1,27 @@
+---
```
[the remark this answers is not in the document]
Comment (ron-ts) [reply docs/resources/user_attributes.md]: Added, 10x

File: docs/resources/user_settings.md [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+<br />
+
+## Example Usage
+
+```terraform
+resource "satori_user_settings" "settings_for_user_json_path" {
+  user_id  = local.user_id_to_configure_settings_on
+
+  // You may define a path to a json file containing a JSON object with the attributes for the user.
+  attributes = "./attributes/user_a.json"
```
Comment (otlg) [inline docs/resources/user_settings.md:23]: I guess you should commit this file as well, isn't it?

File: satori/resources/resource_user_custom_attributes_common.go [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+func validListTypeElement(in []interface{}, t string) bool {
+
+	if !strings.HasPrefix(t, "[]string") &&
+		!strings.HasPrefix(t, "[]int") &&
+		!strings.HasPrefix(t, "[]float") &&
+		!strings.HasPrefix(t, "[]interface") {
+		return false
+	}
+
+	if !strings.HasPrefix(t, "[]interface") {
```
[the remark this answers is not in the document]
Comment (ron-ts) [reply satori/resources/resource_user_custom_attributes_common.go:115]: checking

**Review by loren-tsahalon — APPROVED** [review]

**Pull request merged**